A47-xds-federation.md

```diff
@@ -0,0 +1,682 @@
+  would have been handled prior to this design.
+
+If a gRPC server is created listening on `0.0.0.0:8080`:
+- The existing bootstrap config must already set the server listener
+  resource name template to some non-`xdstp` value, so we'll continue
+  to use that as the resource name.
+- The resource name from template does not start with `xdstp:`, so we
+  do not do any lookup in the `authorities` map.
+- We default to the top-level `xds_server` list, which is the same one
+  that we would have used prior to this design.
+
+#### Example: New-Style Names on gRPC Client
+
+The following new fields are added to the bootstrap config:
+
+```
+"client_default_listener_resource_name_template": "xdstp://xds.example.com/envoy.config.listener.v3.Listener/%s",
+"authorities": {
+  "xds.example.com": {
+  }
+}
+```
+
+If a gRPC client channel is created for `xds:server.example.com`:
+- The target URI specifies no authority, so we use the
+  `client_default_listener_resource_name_template` field.  The
+  resulting resource name will be
+  `xdstp://xds.example.com/envoy.config.listener.v3.Listener/server.example.com`.
+- The resource name specifies the `xds.example.com` authority, so that's
+  the entry we use from the `authorities` map.
+- Note: We do not use the `client_listener_resource_name_template`
+  field in the `authorities` entry, since we've already used the
+  top-level `client_default_listener_resource_name_template` field to
+  construct the resource name.
+- The `xds_server` list is not specified in the entry for the authority,
+  so we default to the top-level `xds_server` list, which is the same one
+  that we would have used prior to this design.
+
+If a gRPC client channel is created for `xds://xds.example.com/server.example.com`:
+- The target URI specifies the authority `xds.example.com`, so that's the
+  entry we use from the `authorities` map.  That entry does not specify any
+  `client_listener_resource_name_template` field, so we use the default of
+  `xdstp://<authority_name>/envoy.config.listener.v3.Listener/%s`.
+  Thus, the resource name will be
+  `xdstp://xds.example.com/envoy.config.listener.v3.Listener/server.example.com`.
+- The `xds_servers` list is not specified in the entry for the authority, so
+  we default to the top-level `xds_servers` list, which is the same one that
+  we would have used prior to this design.
+
+#### Example: New-Style Names on gRPC Server
+
+The following new fields are added to the bootstrap config:
+
+```
+// Use new-style Listener name on gRPC server.
+"server_listener_resource_name_template": "xdstp://xds.example.com/envoy.config.listener.v3.Listener/grpc/server/%s",
+
+// Authorities map.
+"authorities": {
+  "xds.example.com": {
+  }
+}
+```
+
+If a gRPC server is created listening on `0.0.0.0:8080`:
+- The server Listener resource name will be
+  `xdstp://xds.example.com/envoy.config.listener.v3.Listener/grpc/server/0.0.0.0:8080`.
```

Comment:
Should this string have the second `:` percent-encoded?

Reply:
No.  A colon does not need to be percent-encoded in the path part of the URI.

Reply:
I was going by the text on lines 434-438: "... then it will percent-encode the listening address before replacing the %s token in the template." I thought it meant unconditionally percent-encode but you are now saying encode only if needed e.g. if `%s` is in the query parameter part.

Reply:
I don't think colon would need to be percent-encoded anywhere in the URI.  @ejona86 can correct me if I'm wrong.

Reply:
Unconditionally percent-encode would imply the letter 'a' should become %61. `:` is in `pchar` so it can go most places in a URI. That said, it is a bit unclear what should be percent encoded. Should we say to percent encode anything that isn't `unreserved`, `sub-delims`, `:`, `@` ("if it can't be a pchar, without being percent encoded")? That would mean the permitted literal characters are: `ALPHA / DIGIT / "-" / "." / "_" / "~" / "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "=" / ":" / "@"`

Reply:
I meant unconditionally percent-encode reserved characters. In any case it looks okay to not percent-encode ":" in this case based on https://datatracker.ietf.org/doc/html/rfc3986#section-2.2 which has this "If a reserved character is found in a URI component and no delimiting role is known for that character, then it must be interpreted as representing the data octet corresponding to that character's encoding in US-ASCII."